app/assets/javascripts/components/ChallengeModal/ChallengeModal.tsx

```diff
@@ -0,0 +1,242 @@
+  ChallengePrompt,
+  ChallengeReason,
+  ChallengeValue,
+  removeFromArray,
+} from '@standardnotes/snjs';
+import { ProtectedIllustration } from '@standardnotes/stylekit';
+import { FunctionComponent } from 'preact';
+import { useCallback, useEffect, useState } from 'preact/hooks';
+import { Button } from '../Button';
+import { Icon } from '../Icon';
+import { ChallengeModalPrompt } from './ChallengePrompt';
+import { OtherOptionsMenu } from './OtherOptionsMenu';
+
+type InputValue = {
+  prompt: ChallengePrompt;
+  value: string | number | boolean;
+  invalid: boolean;
+};
+
+export type ChallengeModalValues = Record<ChallengePrompt['id'], InputValue>;
+
+type Props = {
+  application: WebApplication;
+  challenge: Challenge;
+  onDismiss: (challenge: Challenge) => Promise<void>;
+};
+
+const validateValues = (
+  values: ChallengeModalValues,
+  prompts: ChallengePrompt[]
+): ChallengeModalValues | undefined => {
+  let hasInvalidValues = false;
+  const validatedValues = { ...values };
+  for (const prompt of prompts) {
+    const value = validatedValues[prompt.id];
+    if (typeof value.value === 'string' && value.value.length === 0) {
+      validatedValues[prompt.id].invalid = true;
+      hasInvalidValues = true;
+    }
+  }
+  if (!hasInvalidValues) {
+    return validatedValues;
+  }
+};
+
+export const ChallengeModal: FunctionComponent<Props> = ({
+  application,
+  challenge,
+  onDismiss,
+}) => {
+  const [values, setValues] = useState<ChallengeModalValues>(() => {
+    const values = {} as ChallengeModalValues;
+    for (const prompt of challenge.prompts) {
+      values[prompt.id] = {
+        prompt,
+        value: prompt.initialValue ?? '',
+        invalid: false,
+      };
+    }
+    return values;
+  });
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [isProcessing, setIsProcessing] = useState(false);
+  const [, setProcessingPrompts] = useState<ChallengePrompt[]>([]);
+  const [bypassModalFocusLock, setBypassModalFocusLock] = useState(false);
+
+  const submit = async () => {
+    const validatedValues = validateValues(values, challenge.prompts);
+    if (!validatedValues) {
+      return;
+    }
+    if (isSubmitting || isProcessing) {
+      return;
+    }
+    setIsSubmitting(true);
+    setIsProcessing(true);
+    const valuesToProcess: ChallengeValue[] = [];
+    for (const inputValue of Object.values(validatedValues)) {
+      const rawValue = inputValue.value;
+      const value = new ChallengeValue(inputValue.prompt, rawValue);
+      valuesToProcess.push(value);
+    }
+    const processingPrompts = valuesToProcess.map((v) => v.prompt);
+    setIsProcessing(processingPrompts.length > 0);
+    setProcessingPrompts(processingPrompts);
+    /**
+     * Unfortunately neccessary to wait 50ms so that the above setState call completely
+     * updates the UI to change processing state, before we enter into UI blocking operation
+     * (crypto key generation)
+     */
+    setTimeout(() => {
```

Comment:
Could we use promises here, which resolves after the above actions? That will let us not to depend from setTimeout and also will eliminate potential issues when 50 ms was not enough (because of heavy system load, for example).